plugins/fpc/fu-fpc-device.c

```diff
@@ -0,0 +1,583 @@
+
+		g_byte_array_append(req, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+
+		if (!fu_fpc_device_dfu_cmd(self,
+					   FPC_CMD_DFU_DNLOAD,
+					   (guint16)i,
+					   req->data,
+					   (gsize)req->len,
+					   FALSE,
+					   FALSE,
+					   &error_local)) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_WRITE,
+				    "failed to write: %s",
+				    error_local->message);
+			return FALSE;
+		}
+
+		if (!fu_device_retry_full(device,
+					  fu_fpc_device_check_dfu_status,
+					  FPC_DFU_MAX_ATTEMPTS,
+					  20,
+					  (gpointer)&dfu_status,
+					  &error_local)) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_WRITE,
+				    "failed to write: %s",
+				    error_local->message);
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_progress_set_percentage_full(fu_progress_get_child(progress),
+						(gsize)i + 1,
+						(gsize)chunks->len);
+	}
+
+	if (!fu_device_has_private_flag(FU_DEVICE(self), FU_FPC_DEVICE_FLAG_LEGACY_DFU)) {
+		/* exit fw download loop. send null package */
+		if (!fu_fpc_device_dfu_cmd(self,
+					   FPC_CMD_DFU_DNLOAD,
+					   0,
+					   NULL,
+					   0,
+					   FALSE,
+					   FALSE,
+					   error)) {
+			g_prefix_error(error, "fail to exit dnload loop:");
+			return FALSE;
+		}
+	}
+	fu_progress_step_done(progress);
+
+	if (!fu_device_retry_full(device,
+				  fu_fpc_device_check_dfu_status,
+				  FPC_DFU_MAX_ATTEMPTS,
+				  20,
+				  (gpointer)&dfu_status,
+				  error))
+		return FALSE;
+
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static void
+fu_fpc_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 2, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 94, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 2, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 2, "reload");
+}
+
+static void
+fu_fpc_device_init(FuFpcDevice *self)
+{
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_USE_RUNTIME_VERSION);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
+	fu_device_set_remove_delay(FU_DEVICE(self), 5000);
+	fu_device_add_protocol(FU_DEVICE(self), "com.fingerprints.dfupc");
+	fu_device_set_summary(FU_DEVICE(self), "FPC fingerprint sensor");
+	fu_device_set_install_duration(FU_DEVICE(self), 15);
+	fu_device_set_firmware_size_min(FU_DEVICE(self), 0x10000);
+	fu_device_set_firmware_size_max(FU_DEVICE(self), 0x50000);
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), FPC_USB_INTERFACE);
+	fu_device_register_private_flag(FU_DEVICE(self),
+					FU_FPC_DEVICE_FLAG_MOH_DEVICE,
+					"moh-device");
+	fu_device_register_private_flag(FU_DEVICE(self), FU_FPC_DEVICE_FLAG_RTS_DEVICE, "rts");
```

Comment:
is `moh-device` and `rts` mutually exclusive? i.e., can you have a device that's bot moh **and** rts?

Reply:
Yes, it's mutually exclusive. Not all moh devices are rts. :)